Dotted line component

erikdstock opened the pull request:
tracked in https://github.com/artsy/emission/issues/677

This PR adds a simple horizontal `DottedLine` component to be used as a separator. Here is the actual aligned with the spec:
![image](https://user-images.githubusercontent.com/9088720/29741828-295ac876-8a42-11e7-8d41-c1c684cc7022.png)

With scaled dots (`scale={5}`):
![image](https://user-images.githubusercontent.com/9088720/29927257-ca1c93f8-8e33-11e7-86eb-79397cebe672.png)
With color (`color={colors['purple-regular']`}
![image](https://user-images.githubusercontent.com/9088720/29927287-e17aaecc-8e33-11e7-887e-99b919308597.png)

I'm not sure if anything is lacking for me to have it display in a conversation- whether i need to log in as a certain user or if these storybook views simply need to be created. **I'm also wondering - does this need a storybook implementation where it displays in context, or is the existing dotted line story sufficient?** 

Thanks!

damassi commented on src/lib/Components/DottedLine.tsx:
```diff
@@ -0,0 +1,14 @@
+import { requireNativeComponent, View } from "react-native"
+
+class DottedLine extends React.Component<any, any> {
+  static propTypes = { dotScale: React.PropTypes.number }
```
Not sure if emission has updated, but `React.PropTypes` has been deprecated and is available in the `prop-types` package: 

```js
import PropTypes from 'prop-types'
```

erikdstock replied:
Ah yes. I'll switch it- thanks.

alloy replied:
Can you add documentation for the `dotScale` prop?

ashfurrow commented on Pod/Classes/DottedLineComponent/ARDottedLineManager.m:
```diff
@@ -0,0 +1,60 @@
+    return line;
+}
+
+@end
```
There're some extra, commented-out code that could probably be removed.

alloy commented on Pod/Classes/DottedLineComponent/ARDottedLineManager.m:
```diff
@@ -0,0 +1,60 @@
+    path.lineWidth = dotDiameter;
+    path.lineCapStyle = kCGLineCapRound;
+    [path moveToPoint:CGPointMake(0, dotDiameter / 2)];
+    [path addLineToPoint:CGPointMake(rect.size.width, 2)];
```
Reading this code it looks like you can control the size of the dots by changing the height of the `DottedLine` component. If correct, can you add a second story that demonstrates this?

alloy commented on src/lib/Components/DottedLine.tsx:
```diff
@@ -0,0 +1,14 @@
+import * as React from "react"
+import { requireNativeComponent, View } from "react-native"
+
+class DottedLine extends React.Component<any, any> {
```
Can you add a TypeScript interface for the props that describes the `dotScale` prop? And specify that the state will be `null`?

orta replied:
( there's some docs on how that works here: https://github.com/artsy/reaction/blob/master/docs/getting_started.md )

erikdstock replied:
yep, thank you!

alloy commented on src/lib/Components/__stories__/DottedLine.story.tsx:
```diff
@@ -0,0 +1,15 @@
+      {story()}
+    </View>
+  )
+  .add("Flat White", () => {
```
I think this title is the result from some copy-pasta, right? Can you name this something like ‘With default props’?

erikdstock replied:
Oh man. I kept looking at that name and thinking 'why did we name it this when it is neither flat nor white?'

alloy replied:
😂

erikdstock commented:
Alright, I was able to come back to this and am getting close to finishing up but am not sure about how this should work: basically the container component in the storybook does not constrain or set bounds for the dotted line inside it- if the dots are scaled larger they will expand outside the container. Here i have set the height of that 'wrapperView' artificially low to 2. 

My goal was to have some basic color and scale controls, but I'm also wondering if this isn't worth it.

![image](https://user-images.githubusercontent.com/9088720/29795752-954c1f8c-8c1b-11e7-9ef7-3133eb1367c2.png)

mennenia commented:
Amazing @erikdstock!!! Looks GREAT 😍

alloy commented:
@erikdstock I think the component should internally default to the smallest height that looks good (as per the design), but you can definitely expose that property as something that the user can control. To be clear, that means that the `DottedLine` wrapper component would set that height on the `NativeDottedLine`. Does that answer your question?

erikdstock commented:
@alloy I believe that is what I'm doing right now. The specific issue pictured above is that the story above has a wrapper (unrelated to the DottedLine component itself) and that wrapper does not affect the line. Since it is a line, maybe it shouldn't matter. But to clarify, besides the very low wrapperView height I set above when I was figuring this out, this is the setup you want?
```jsx
// story
    <View accessibilityLabel="wrapperView" style={{ height: 2, marginTop: 60, marginLeft: 20, marginRight: 20 }}>
      {story()}
    </View>

// DottedLine render()
    return <NativeDottedLine style={{ height: this.props.scale * 2 }} color={dotColor} />
```

erikdstock commented:
Updated with documentation- but I still can't see it in context of the `Conversation`. Is that storybook broken, or can I see it when I am logged in as a different user?

If we don't need that, I removed the WIP line above.

alloy commented on Pod/Classes/DottedLineComponent/ARDottedLineManager.m:
```diff
@@ -0,0 +1,55 @@
+{
+    [super drawRect:rect];
+    
+    [[UIColor whiteColor] setFill];
```
I wonder if this should be `clearColor` instead and that’ll show the white background underneath. If it doesn’t then this is fine too 👍

alloy commented:
When used in the conversation and inspecting the view hierarchy with Xcode’s view debugger, do you see it exist at all and what are its dimensions?

erikdstock commented:
Had a chanc3e to revisit today- showing as just the height of the dots.
However, these were initially rending with `clearColor()` on a black background, so I set it up with a custom `backgroundColor` prop. In doing that I noticed some XCode documentation pop up saying that backgroundColor defaults to `nil`/transparent so to avoid possibly colliding with something I changed the name to `processedBackgroundColor`. Then it turned black, so I made it white by default. `backgroundColor={null}` in props will achieve the `UIColor *nil` black effect, but I don't know if that is my error or the actual color of a transparent view 🤔. 
![image](https://user-images.githubusercontent.com/9088720/30005705-6ad78a26-90b6-11e7-8646-af9f42fc6282.png)

Actual view:
![image](https://user-images.githubusercontent.com/9088720/30005727-c0ff6edc-90b6-11e7-8a8a-b6f281040a0a.png)

alloy commented:
Yeah instead of `nil` you should have it default to `[UIColor clearColor]` and let it draw that, by default the graphics context is going to be black.

damassi commented:
One little design nit -- maybe increase the line padding between each message item so that the margins between appear evenly spaced? Right now the line is much closer to "Karl and Anna" then it is to "ACA Galaries" (or the thumbnail image) which throws off the overall design.

erikdstock commented:
Updated with @orta to fix the margin in context and remove height/color options- will always now be `gray-medium` on white with a scale to match the spec:
![image](https://user-images.githubusercontent.com/9088720/30174777-30f7688c-93ca-11e7-8d8d-51b345b9bfac.png)

erikdstock commented:
Closing to open a new pr directly from the upstream.
